File: go/bob/bob.go
// Package bob provides a Hey method for responding as a lackadaisical teeneager
package bob

import (
	"regexp"
	"strings"
)

// Hey returns a teenager's response based on the original response
func Hey(remark string) string {
	trimmedRemark := strings.Trim(remark, " \t\n\r")
	isQuestion := strings.HasSuffix(trimmedRemark, "?")
	hasLetters, _ := regexp.MatchString("[A-z]", trimmedRemark)

	switch {
	case trimmedRemark == "":
		return "Fine. Be that way!"
	case hasLetters && strings.ToUpper(trimmedRemark) == trimmedRemark:
		if isQuestion {
			return "Calm down, I know what I'm doing!"
		}
		return "Whoa, chill out!"
	case isQuestion:
		return "Sure."
	default:
		return "Whatever."
	}
}

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      String.trim(input) == "" ->
        "Fine. Be that way!"

      question?(input) && shouting?(input) ->
        "Calm down, I know what I'm doing!"

      question?(input) ->
        "Sure."

      shouting?(input) ->
        "Whoa, chill out!"

      true ->
        "Whatever."
    end
  end

  defp shouting?(input) do
    String.upcase(input) == input &&
      String.downcase(input) != input
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end
end
